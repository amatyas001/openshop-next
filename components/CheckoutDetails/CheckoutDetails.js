import { v1 as uuidv1 } from 'uuid';
import Router, { useRouter } from 'next/router';
import { useStripe } from '@stripe/react-stripe-js';
import { useSelector, useDispatch } from 'react-redux';
import {
  paymentReset,
  checkoutFinished,
  paymentToken,
  clearCart,
} from '../../redux/actions';
import { Flex, Text, Heading, Spinner } from '@chakra-ui/core';
import { StripeForm, StripeConfirm, Button } from '../index';

export const CheckoutDetails = () => {
  const stripe = useStripe();
  const dispatch = useDispatch();
  const payment = useSelector((state) => state.payment);
  const router = useRouter();
  const [content, setContent] = React.useState(false);

  React.useEffect(() => {
    Router.prefetch('/');
  }, []);

  React.useEffect(() => {
    if (payment) {
      if (!payment.status) router.replace('/');
      dispatch(paymentToken(uuidv1()));
      switch (payment.status) {
        case 'success':
          dispatch(checkoutFinished());
          setContent(
            <>
              <Heading>Payment Successful</Heading>
              <Text fontSize='1.5rem'>
                Received intent number:{' '}
                <Text as='span' fontWeight='bold' color='purple.800'>
                  {payment.intent}
                </Text>
              </Text>
              <Text>
                This was a sample transaction to demonstrate the payment flow.
                Thank you!
              </Text>
            </>
          );
          break;
        case 'canceled':
          dispatch(clearCart());
          dispatch(checkoutFinished());
          setContent(
            <>
              <Heading>Payment Cancelled</Heading>
              <Text fontSize='1.5rem'>
                Successfuly cancelled intent:{' '}
                <Text as='span' fontWeight='bold' color='purple.800'>
                  {payment.intent.id}
                </Text>
              </Text>
            </>
          );
          break;
        case 'error':
          setContent(
            <>
              <Text
                py='15px'
                px='20px'
                my='50px'
                fontSize='1.3rem'
                bg='red.400'
                color='gray.100'
                borderRadius='5px'
                d='block'
                width='100%'
              >
                <strong>Error:</strong>
                <br />
                <em>{payment.error}</em>
              </Text>
              <Button onClick={() => dispatch(paymentReset())}>retry</Button>
            </>
          );
          break;
        default:
          setContent(<Spinner size='150px' size='150px' my='50px' />);
      }
    }
  }, [payment, stripe]);

  return (
    <Flex
      flexDirection='column'
      alignItems='center'
      width={{ sm: '100%', md: '80%', lg: '60%', xl: '50%' }}
      mx='auto'
    >
      {payment.status !== 'success' &&
      payment.status !== 'error' &&
      payment.status !== 'canceled' ? (
        <>
          <StripeForm />
          <StripeConfirm />
        </>
      ) : (
        content || <Spinner size='150px' size='150px' my='50px' />
      )}
    </Flex>
  );
};
