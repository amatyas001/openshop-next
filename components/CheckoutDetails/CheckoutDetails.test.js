import { Provider } from 'react-redux';
import { create } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import {
  paymentReset,
  checkoutFinished,
  paymentToken,
  clearCart,
} from '../../redux/actions';
import { CheckoutDetails } from './CheckoutDetails';

const mockStore = configureStore([]);
const mock_prefetch = jest.fn();
const mock_replace = jest.fn();
const mock_uuid = 'uuid';

jest.mock('@stripe/react-stripe-js', () => ({
  useStripe: () => true,
  useElements: jest.fn(),
  CardElement: () => 'CardElement',
}));

jest.mock('next/router', () => ({
  useRouter: () => ({
    replace: mock_replace,
    prefetch: mock_prefetch,
  }),
}));

jest.mock('uuid', () => ({
  v1: () => mock_uuid,
}));

describe('<CheckoutDetails />', () => {
  let tree;

  afterEach(() => {
    mock_prefetch.mockClear();
    mock_replace.mockClear();
  });

  describe('when payment initialized', () => {
    let store;
    beforeEach(() => {
      store = mockStore({
        payment: { status: true },
      });
      tree = create(
        <Provider store={store}>
          <CheckoutDetails />
        </Provider>
      );
    });

    it('should render without props', () => {
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should prefetch index page', () => {
      expect(mock_prefetch).toHaveBeenCalledTimes(1);
      expect(mock_prefetch).toHaveBeenCalledWith('/');
    });

    it('should create token', () => {
      expect(store.getActions()).toEqual([paymentToken(mock_uuid)]);
    });

    describe('on form incomplete', () => {
      it('should render details form', () => {
        expect(tree.root.findByType('form').type).toBe('form');
      });
    });

    describe('on form complete', () => {
      const states = ['success', 'canceled', 'error'];
      states.map((item) => {
        it(`should render ${item} content`, () => {
          store = mockStore({
            payment: { status: item, intent: { id: 'id' } },
          });
          tree = create(
            <Provider store={store}>
              <CheckoutDetails />
            </Provider>
          );
          switch (item) {
            case 'success':
              expect(store.getActions()).toEqual([
                paymentToken(mock_uuid),
                checkoutFinished(),
              ]);
              break;
            case 'canceled':
              expect(store.getActions()).toEqual([
                paymentToken(mock_uuid),
                clearCart(),
                checkoutFinished(),
              ]);
              break;
            case 'error':
              expect(store.getActions()).toEqual([paymentToken(mock_uuid)]);
              break;
          }
          expect(tree.toJSON()).toMatchSnapshot();
        });
      });
    });

    describe('on error', () => {
      it('should reset progress', () => {
        store = mockStore({
          payment: { status: 'error' },
        });
        tree = create(
          <Provider store={store}>
            <CheckoutDetails />
          </Provider>
        );
        tree.root.findByType('button').props.onClick();
        expect(store.getActions()).toEqual([
          paymentToken(mock_uuid),
          paymentReset(),
        ]);
      });
    });
  });

  describe('when payment not initialized', () => {
    let store;
    beforeEach(() => {
      store = mockStore({
        payment: { status: false },
      });
      tree = create(
        <Provider store={store}>
          <CheckoutDetails />
        </Provider>
      );
    });

    it('should render without props', () => {
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should redirect to home page', () => {
      expect(mock_replace).toHaveBeenCalledTimes(1);
      expect(mock_replace).toHaveBeenCalledWith('/');
    });
  });
});
