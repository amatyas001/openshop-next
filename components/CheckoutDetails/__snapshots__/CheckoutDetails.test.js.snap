// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckoutDetails /> when payment initialized on form complete should render canceled content 1`] = `
<div
  className="css-z4bb15"
>
  <h2
    className="css-84al3w"
  >
    Payment Cancelled
  </h2>
  <p
    className="css-441ftm"
  >
    Successfuly cancelled intent:
     
    <span
      className="css-4i139l"
    >
      id
    </span>
  </p>
</div>
`;

exports[`<CheckoutDetails /> when payment initialized on form complete should render error content 1`] = `
<div
  className="css-z4bb15"
>
  <p
    className="css-1po2p9s"
  >
    <strong>
      Error:
    </strong>
    <br />
    <em />
  </p>
  <button
    className="css-1729dnv"
    data-cy="button"
    href="#0"
    onClick={[Function]}
    onKeyDown={[Function]}
    tabIndex={0}
  >
    retry
  </button>
</div>
`;

exports[`<CheckoutDetails /> when payment initialized on form complete should render success content 1`] = `
<div
  className="css-z4bb15"
>
  <h2
    className="css-84al3w"
  >
    Payment Successful
  </h2>
  <p
    className="css-441ftm"
  >
    Received intent number:
     
    <span
      className="css-4i139l"
    >
      id
    </span>
  </p>
  <p
    className="css-18poma1"
  >
    This was a sample transaction to demonstrate the payment flow. Thank you!
  </p>
</div>
`;

exports[`<CheckoutDetails /> when payment initialized should render without props 1`] = `
<div
  className="css-z4bb15"
>
  <form
    className="css-1mi3tt8"
  >
    <fieldset
      className="css-94sf3x"
    >
      <fieldset
        className="css-9x3bhi"
      >
        <label
          className="css-1xnp38i"
          htmlFor="name"
        >
          name
        </label>
        <input
          autoComplete="name"
          className="css-cth1sm"
          id="name"
          onChange={[Function]}
          placeholder="Jane Doe"
          required={true}
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        className="css-9x3bhi"
      >
        <label
          className="css-1xnp38i"
          htmlFor="address"
        >
          address
        </label>
        <input
          autoComplete="address"
          className="css-cth1sm"
          id="address"
          onChange={[Function]}
          placeholder="J191 S. Illinois Street, Everett, MA 02149"
          required={true}
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        className="css-9x3bhi"
      >
        <label
          className="css-1xnp38i"
          htmlFor="email"
        >
          email
        </label>
        <input
          autoComplete="email"
          className="css-cth1sm"
          id="email"
          onChange={[Function]}
          placeholder="janedoe@gmail.com"
          required={true}
          type="email"
          value=""
        />
      </fieldset>
      <fieldset
        className="css-9x3bhi"
      >
        <label
          className="css-1xnp38i"
          htmlFor="phone"
        >
          phone
        </label>
        <input
          autoComplete="tel"
          className="css-cth1sm"
          id="phone"
          onChange={[Function]}
          placeholder="(941) 555-0123"
          required={true}
          type="tel"
          value=""
        />
      </fieldset>
      <div
        className="css-70qvj9"
      >
        <label
          className="css-1gjn3mk"
          htmlFor="StripeElement"
        >
          card
        </label>
        <p
          className="css-1i3anel"
        >
          Use any date and CVC with 
          <strong>
            4242 4242 4242 4242
          </strong>
           
          number for testing
        </p>
        <p
          className="css-cz9gy2"
        />
      </div>
      CardElement
    </fieldset>
    <p
      className="css-1ebbm78"
    >
      By clicking the
       
      <span
        className="css-1yhuerr"
      >
        PAY
      </span>
       
      button you will accept the
       
      <a
        className="css-1c5zfe9"
      >
        Terms of Service
      </a>
       
      of the company.
    </p>
    <button
      className="css-2jdihq"
      data-cy="button"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      BACK TO REVIEW
    </button>
    <button
      className="css-u7u20l"
      data-cy="button"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <span
        className="css-1d22953"
      >
        FILL THE FORM
      </span>
    </button>
  </form>
</div>
`;

exports[`<CheckoutDetails /> when payment not initialized should render without props 1`] = `
<div
  className="css-z4bb15"
>
  <form
    className="css-1mi3tt8"
  >
    <fieldset
      className="css-94sf3x"
    >
      <fieldset
        className="css-9x3bhi"
      >
        <label
          className="css-1xnp38i"
          htmlFor="name"
        >
          name
        </label>
        <input
          autoComplete="name"
          className="css-cth1sm"
          id="name"
          onChange={[Function]}
          placeholder="Jane Doe"
          required={true}
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        className="css-9x3bhi"
      >
        <label
          className="css-1xnp38i"
          htmlFor="address"
        >
          address
        </label>
        <input
          autoComplete="address"
          className="css-cth1sm"
          id="address"
          onChange={[Function]}
          placeholder="J191 S. Illinois Street, Everett, MA 02149"
          required={true}
          type="text"
          value=""
        />
      </fieldset>
      <fieldset
        className="css-9x3bhi"
      >
        <label
          className="css-1xnp38i"
          htmlFor="email"
        >
          email
        </label>
        <input
          autoComplete="email"
          className="css-cth1sm"
          id="email"
          onChange={[Function]}
          placeholder="janedoe@gmail.com"
          required={true}
          type="email"
          value=""
        />
      </fieldset>
      <fieldset
        className="css-9x3bhi"
      >
        <label
          className="css-1xnp38i"
          htmlFor="phone"
        >
          phone
        </label>
        <input
          autoComplete="tel"
          className="css-cth1sm"
          id="phone"
          onChange={[Function]}
          placeholder="(941) 555-0123"
          required={true}
          type="tel"
          value=""
        />
      </fieldset>
      <div
        className="css-70qvj9"
      >
        <label
          className="css-1gjn3mk"
          htmlFor="StripeElement"
        >
          card
        </label>
        <p
          className="css-1i3anel"
        >
          Use any date and CVC with 
          <strong>
            4242 4242 4242 4242
          </strong>
           
          number for testing
        </p>
        <p
          className="css-cz9gy2"
        />
      </div>
      CardElement
    </fieldset>
    <p
      className="css-1ebbm78"
    >
      By clicking the
       
      <span
        className="css-1yhuerr"
      >
        PAY
      </span>
       
      button you will accept the
       
      <a
        className="css-1c5zfe9"
      >
        Terms of Service
      </a>
       
      of the company.
    </p>
    <button
      className="css-2jdihq"
      data-cy="button"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      BACK TO REVIEW
    </button>
    <button
      className="css-u7u20l"
      data-cy="button"
      href="#0"
      onClick={[Function]}
      onKeyDown={[Function]}
      tabIndex={0}
    >
      <span
        className="css-1d22953"
      >
        FILL THE FORM
      </span>
    </button>
  </form>
</div>
`;
